### PR TITLE
Remove pydantic from DJ client.

### PR DESCRIPTION
**Pull request by agorajek**
### Summary

Remove pydantic from the DJ client to lower the dependency hell for DJ users.

Some pydantic functionality (like validate and simple attributes) are replaced with data classes.

### Test Plan

Unit tests for now. Need to do more.

- [x] PR has an associated issue: #1170
- [x] `make check` passes
- [x] `make test` shows 100% unit test coverage

### Deployment Plan

deploy and profit

### CODE REVIEW

File: datajunction-clients/python/tests/examples/project1/roads/repair_orders_cube.cube.yaml [hunk 1]
```diff
@@ -6,4 +6,4 @@ metrics:
 dimensions:
   - ${prefix}roads.hard_hat.state
   - ${prefix}roads.dispatcher.company_name
-filter: []
+filters: []
```
Comment (agorajek) [inline datajunction-clients/python/tests/examples/project1/roads/repair_orders_cube.cube.yaml:9]: Hey @samredai check this out... it turns out dataclasses do a better job in attribute validation than pydantic ;)

File: datajunction-clients/python/datajunction/builder.py [hunk 2]
```diff
@@ -238,8 +283,21 @@ def register_table(self, catalog: str, schema: str, table: str) -> Source:
                 f"Failed to register table `{catalog}.{schema}.{table}`: {exc}",
             ) from exc
         source_node = Source(
-            **response.json(),
             dj_client=self,
+            name=response.json()["name"],
```
Comment (samredai) [inline datajunction-clients/python/datajunction/builder.py:287]: Might be better to serialize to json once, like `data = response.json()` then do `data["name"]` to pull out the values.

File: datajunction-clients/python/datajunction/builder.py [hunk 3]
```diff
@@ -86,25 +87,67 @@ def restore_namespace(self, namespace: str, cascade: bool = False) -> None:
...
+        raise DJClientException(f"Unknown node type: {type_}")  # pragma: no cover
+
     def create_node(
         self,
         type_: models.NodeType,
         name: str,
         data: Dict,
         update_if_exists: bool = True,
-    ) -> "Node":
+    ):
```
Comment (samredai) [inline datajunction-clients/python/datajunction/builder.py:146]: Why'd this return type get removed?

File: datajunction-clients/python/datajunction/_internal.py [hunk 4]
```diff
@@ -582,7 +588,7 @@ def _create_tag(
         response = self._session.post(
             "/tags/",
             timeout=self._timeout,
-            json=tag.dict(exclude_none=True),
+            json=tag.to_dict(),
```
Comment (samredai) [inline datajunction-clients/python/datajunction/_internal.py:591]: Is there a reason why we use a `to_dict()` method sometimes and `asdict(...)` sometimes? I believe asdict would also convert any nested dataclasses.

File: datajunction-clients/python/datajunction/nodes.py [hunk 5]
```diff
@@ -58,24 +63,61 @@ def cubes(self):
...
+        dict_ = {
+            "name": self.name,
+            "type": self.type,
+            "description": self.description,
+            "mode": self.mode,
+            "status": self.status,
+            "display_name": self.display_name,
+            "availability": self.availability,
+            "tags": [
+                tag.to_dict() if isinstance(tag, Tag) else tag for tag in self.tags
```
Comment (samredai) [inline datajunction-clients/python/datajunction/nodes.py:101]: This is a case of nesting where I think `dataclasses.asdict(...)` would handle on its own.

File: datajunction-clients/python/pyproject.toml [hunk 6]
```diff
@@ -11,9 +11,7 @@ authors = [
 ]
 dependencies = [
     "requests<3.0.0,>=2.28.2",
-    "pydantic>=1.10.7,<2",
```
Comment (samredai) [inline datajunction-clients/python/pyproject.toml:14]: image

File: datajunction-clients/python/tests/examples/project1/roads/repair_orders_cube.cube.yaml (hunk 1 above)
Comment (samredai) [reply datajunction-clients/python/tests/examples/project1/roads/repair_orders_cube.cube.yaml:9]: Wow, look at that! 🤯😭

**Review by samredai — APPROVED** [review]
This is awesome @agorajek, thank you! I left a number of comments but overall LGTM. 👍

File: datajunction-clients/python/datajunction/builder.py [hunk 7]
```diff
@@ -86,25 +87,67 @@ def restore_namespace(self, namespace: str, cascade: bool = False) -> None:
...
+        """
+        Make a new node of the given type.
+        """
+        # common arguments
+        common_args = [
+            field.name
+            for field in fields(Node)
+            if field.name not in ["dj_client", "type"]
+        ]
+        if type_ == models.NodeType.SOURCE:
```
Comment (shangyian) [inline datajunction-clients/python/datajunction/builder.py:104]: You can probably reduce these to a single call?

File: datajunction-clients/python/datajunction/nodes.py [hunk 8]
```diff
@@ -276,31 +327,62 @@ def set_column_attributes(
...
     type: str = "source"
-    catalog: str
-    schema_: str
-    table: str
-    columns: Optional[List[models.Column]]
+    catalog: Optional[str] = None
+    schema_: Optional[str] = None
+    table: Optional[str] = None
+
+    def to_dict(self, exclude: Optional[List[str]] = None) -> Dict[str, Any]:
```
Comment (shangyian) [inline datajunction-clients/python/datajunction/nodes.py:341]: I think calls to this function could be replaced by `dataclasses.asdict`, no?

**Review by shangyian — APPROVED** [review]
Yay!!!

File: datajunction-clients/python/datajunction/_internal.py (hunk 4 above)
Comment (agorajek) [reply datajunction-clients/python/datajunction/_internal.py:591]: No reason other than the PR took me a while and I was going back and forth between solutions ;)

File: datajunction-clients/python/datajunction/builder.py (hunk 7 above)
Comment (agorajek) [reply datajunction-clients/python/datajunction/builder.py:104]: Ah, you are right. I probably can.

File: datajunction-clients/python/datajunction/nodes.py (hunk 8 above)
Comment (agorajek) [reply datajunction-clients/python/datajunction/nodes.py:341]: Yes...

File: datajunction-clients/python/datajunction/_internal.py (hunk 4 above)
Comment (agorajek) [reply datajunction-clients/python/datajunction/_internal.py:591]: Actually there is one more I forgot to mention.... dataclasses asdict() method does not handle nested dataclasses well:
```
!!! Recursion detected (same locals & position)
```

File: datajunction-clients/python/datajunction/nodes.py (hunk 5 above)
Comment (agorajek) [reply datajunction-clients/python/datajunction/nodes.py:101]: Yes, it fails on nested objects.

File: datajunction-clients/python/datajunction/nodes.py (hunk 8 above)
Comment (agorajek) [reply datajunction-clients/python/datajunction/nodes.py:341]: Actually no. Dataclasses fail miserably on `asdict()` when they have any nested dataclasses. That's why I needed this to_dict().

**Pull request merged**